### ISSUE TITLE
Unable to statically link xmlsec in MSYS2 environment. Undefined reference.

### ISSUE THREAD

**Issue by totszwai** [lsh123/xmlsec#675]
We used to run with 1.2.31, everything compiled fine. Switched to 1.2.37 and everything starts throwing these errors.
Has xmlSecError and a few others calls been removed?

![image](https://github.com/lsh123/xmlsec/assets/4440365/44716ace-b1e7-414e-a8a4-c5f8d06d5b1a)

**Comment by totszwai** [comment]
I think we did rebuild the whole thing against the 1.2.37, I will go try again, but is there any special flag we should be using if we are still on OpenSSL1.1.1o? We are not on OpenSSL v3 yet.

Thank you for the fast reply, meanwhile, I will wipe the whole build and make sure everything is on 1.2.37.

**Comment by lsh123** [comment]
It should detect the openssl version automatically. You can see it in the ./configure output. Regardless this error doesn't look related to OpenSSL 1.1.x vs 3.x

One idea is that the system xmlsec libraries get picked up for some reasons. You might want to check what paths are used during a) compiling; b) linking; c) running. Currently it looks like a) and b) use different "flavors" of the library

**Comment by totszwai** [comment]
@lsh123 I made sure they are all using the same version. Actually, I don't think it is related to OpenSSL either, since the error messages (a bit weird) is about libxmlsec not able to find those reference?

![image](https://github.com/lsh123/xmlsec/assets/4440365/b5538587-d7dc-4be7-beaa-300557431901)

But it looks like is there...
```
$ grep -r "__imp_xmlSecErrorsGetCode" *
Binary file src/xmlsec1-1.2.37/src/.libs/libxmlsec1.dll.a matches
Binary file src/xmlsec1-1.2.37/src/openssl/.libs/libxmlsec1-openssl.a matches
Binary file src/xmlsec1-1.2.37/src/openssl/.libs/libxmlsec1-openssl.dll matches
Binary file src/xmlsec1-1.2.37/src/openssl/.libs/libxmlsec1_openssl_la-crypto.o matches
Binary file src/xmlsec1-1.2.37/src/openssl/libxmlsec1_openssl_la-crypto.o matches
```

This is (more or less) the flags we are compiling with:

```
x86_64-w64-mingw32-g++ -fstack-protector-all  my-app.o libcislsapval-static.a hashcheck/libhashcheck-static.a libcislsapgui-static.a valreport.o -pthread -lgtk-3 -latk-1.0 -lgio-2.0 -lgdk-3 -lz -lgdi32 -limm32 -lshell32 -lole32 -luuid -lwinmm -ldwmapi -lsetupapi -lcfgmgr32 -lhid -lwinspool -lcomctl32 -lcomdlg32 -lpangowin32-1.0 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgobject-2.0 -lgmodule-2.0 -lglib-2.0 -lintl  -lmincore -lgdiplus -Wl,--export-all-symbols -L/c/Users/et5341/source/windows/libcarillon/src -lcarillon-static -L/c/Users/et5341/source/windows/xmltools/src -lxmltools-static -L/c/Users/et5341/source/windows/tss/src -ltsv /c/Users/et5341/source/windows/xmlsec-custom/src/xmlsec1-1.2.37/src/openssl/.libs/libxmlsec1-openssl.a /c/Users/et5341/source/windows/xmlsec-custom/src/xmlsec1-1.2.37/src/.libs/libxmlsec1.a /c/Users/et5341/source/windows/libxslt-custom/src/libxslt-1.1.34/libxslt/.libs/libxslt.a /c/Users/et5341/source/windows/libxml2-custom/src/libxml2-2.9.13/.libs/libxml2.a -largp  /c/Users/et5341/source/windows/openssl-custom/src/openssl-1.1.1o/libssl.a /c/Users/et5341/source/windows/openssl-custom/src/openssl-1.1.1o/libcrypto.a /c/Users/et5341/source/windows/zlib-custom/src/zlib-1.2.13/libz.a   -pthread -lgtk-3 -latk-1.0 -lgio-2.0 -lgdk-3 -lz -lgdi32 -limm32 -lshell32 -lole32 -luuid -lwinmm -ldwmapi -lsetupapi -lcfgmgr32 -lhid -lwinspool -lcomctl32 -lcomdlg32 -lpangowin32-1.0 -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgobject-2.0 -lgmodule-2.0 -lglib-2.0 -lintl  -lmincore -lgdiplus -Wl,--export-all-symbols -o my-app.exe
```


I also tried to specify the `--with-openssl` manually, and the result is exactly the same. Any other recommendation on how I could troubleshoot this? Thanks!

**Comment by totszwai** [comment]
@lsh123 what changed?

Comparing the output of both version...  interestingly enough, the old version in their libxmlsec1-openssl.a and libxmlsec1.a, they never had any of those definitions in them, it is only in the version 37.

![grep-1 2 31](https://github.com/lsh123/xmlsec/assets/4440365/6eea812f-6452-460d-9299-ae59cac2ace8)

![grep-1 2 37](https://github.com/lsh123/xmlsec/assets/4440365/4b964d55-10cf-42d0-955e-a5f33e5c16a9)

**Comment by lsh123** [comment]
No idea. Something is different in how you build the library.

**Comment by totszwai** [comment]
@lsh123 I can confirm that version 1.2.31 (we started with this version), 32, 33, 34 all worked just fine. It only starts throwing those errors starting at 35.

**Comment by totszwai** [comment]
@lsh123 I can confirm that version 1.2.31 (we started with this version), 32, 33, 34 all worked just fine. It only starts throwing those errors starting at 35.

**Comment by lsh123** [comment]
Even in your environment, the xmlsec command line app was built correctly. It’s very hard for me to help since there is no repro test case I can look into.

**Comment by lsh123** [comment]
Actually, can you try to build xmlsec examples in the examples/ folder?

**Comment by totszwai** [comment]
@lsh123 Same problem, we are no longer able to statically linking them.

The only way I could link it successfully, was to link them against the `dll.a` instead of just `.a` file. 
Note: Previously, we would just link the `.a` with our Windows executables and would work.

![image](https://github.com/lsh123/xmlsec/assets/4440365/a338954e-d928-4194-961c-0cfd433d7132)

To get the 1.2.37 statically link with our executables, I changed `.a` to `.dll.a` however, at runtime it would still ended up asking for the dll.

![image](https://github.com/lsh123/xmlsec/assets/4440365/40be8ba7-0f0a-4568-bc4c-701da4befa1d)
![image](https://github.com/lsh123/xmlsec/assets/4440365/6ca516d7-7644-47d5-8673-58b43cc7b436)


I am able to build all the example statically linking them (with `.dll.a`) but running the examples as Windows executables is a problem.
![image](https://github.com/lsh123/xmlsec/assets/4440365/d85b3adb-68e8-4222-9cb7-f00447795474)

Any advise?

**Comment by lsh123** [comment]
OK, it's a bit puzzling because xmlsec command line utility compiles and runs (`make check` passes) on CI. I will take a look but likely it will be after July 4th.

Re errors with DLLs, try to copy dlls instead of linking.

**Comment by totszwai** [comment]
@lsh123 I can build and run the command line but only linking them and using the dll. However, we don't want to use the dlls.
![image](https://github.com/lsh123/xmlsec/assets/4440365/dc63f904-c860-4504-b11c-2ae6ba4338f3)

Also, in the previous reply, I had the below error, but is resolved after I bring over the libxslt-1.dll (discovered that after I got the `make check` to work):
![image](https://github.com/lsh123/xmlsec/assets/4440365/68e00956-5ed6-452f-bfaa-46bd4c44fba3)


Just to reiterate, we can't link them against the `.a` file, so far, it only works with `.dll.a` **AND** bring over all the related dlls.

This is from those examples:
```
x86_64-w64-mingw32-gcc -Wall -fstack-protector-all -fpie -DDSO_WIN32 -DWIN32_LEAN_AND_MEAN -m64 -Wno-unknown-pragmas   -O2 -I/mingw64/include -I/usr/x86_64-w64-mingw32/include -I/c/Users/et5341/source/windows/xmlsec-custom/src/xmlsec1-1.2.37/include -DXMLSEC_CRYPTO_OPENSSL=1 -DXMLSEC_NO_SIZE_T -I/c/Users/et5341/source/windows/libxml2-custom/src/libxml2-2.9.13/include -I/c/Users/et5341/source/windows/libxslt-custom/src/libxslt-1.1.34 -fstack-protector-all   sign1.c  /c/Users/et5341/source/windows/xmlsec-custom/src/xmlsec1-1.2.37/src/openssl/.libs/libxmlsec1-openssl.a /c/Users/et5341/source/windows/xmlsec-custom/src/xmlsec1-1.2.37/src/.libs/libxmlsec1.a /c/Users/et5341/source/windows/libxslt-custom/src/libxslt-1.1.34/libxslt/.libs/libxslt.a /c/Users/et5341/source/windows/libxml2-custom/src/libxml2-2.9.13/.libs/libxml2.a -o sign1
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x86): undefined reference to `__imp_xmlParseFile'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x99): undefined reference to `__imp_xmlDocGetRootElement'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0xb3): undefined reference to `__imp_xmlSecDSigNs'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0xba): undefined reference to `__imp_xmlSecNodeSignature'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0xc3): undefined reference to `__imp_xmlSecFindNode'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0xd7): undefined reference to `__imp_xmlSecDSigCtxCreate'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x100): undefined reference to `__imp_xmlSecOpenSSLAppKeyLoad'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x11c): undefined reference to `__imp_xmlSecKeySetName'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x130): undefined reference to `__imp_xmlSecDSigCtxSign'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x151): undefined reference to `__imp_xmlDocDump'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x15a): undefined reference to `__imp_xmlSecDSigCtxDestroy'
C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/13.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: C:\msys64\tmp\ccSPWETi.o:sign1.c:(.text+0x163): undefined reference to `__imp_xmlFreeDoc'
```

**Comment by lsh123** [comment]
The fact that “DLL-version” works gives you a strong hint that all symbols, functions, etc are there and as I was saying it’s purely a build issue. The fact that the linker looks for “__imp_foobar” symbols is suspicious -  I believe this prefix is used fir dlls:

https://stackoverflow.com/questions/5159353/how-can-i-get-rid-of-the-imp-prefix-in-the-linker-in-vc


Check if you have XMLSEC_STATIC defined when you compile your code - it controls how exactly functions are imported / exported on Windows:

https://github.com/lsh123/xmlsec/blob/master/include/xmlsec/exports.h

**Comment by lsh123** [comment]
BTW, if it is indeed the issue, I bet this diff is the reason and it was working before by mistake:

https://github.com/lsh123/xmlsec/commit/4d6de9af4c4dc616b3c8c4fc80c9a0a9e51c920e#diff-54d4c81f073b275546b7206c59bce4dfa89e16cbf28b2b43847ccc47b9917cd5

**Comment by lsh123** [comment]
Sorry finally got home and was able to look into this in more details. I can reproduce the issue and I have no idea how it worked before. Unfortunately on Windows you can't compile both dynamically and statically linked libraries in the same time because the function names are different. So you have to choose one or another.  I fixed the problem on master (https://github.com/lsh123/xmlsec/pull/679) and on the 1.2.x branch (https://github.com/lsh123/xmlsec/commit/6403818687eb3e10808fe60af694e509a2a2279c). After these fixes, if you want to compile statically on MinGW / Cygwin then you need to build xmlsec as follows:

```
./configure --enable-static  --enable-static-linking <other options as needed>
make
```

And then compile your code with XMLSEC_STATIC=1 defined. 

I am also adding the static MinGW build to the CI so this is not getting broken by mistake (https://github.com/lsh123/xmlsec/pull/678). 

Thanks for bug report and sorry it took so many back-and-froths as I was traveling.

**Comment by totszwai** [comment]
@lsh123 would it be possible to generate a release? I downloaded the latest, ran `autoreconf -i` then try to build it but getting:

```
...
./configure: line 11032: xmlsec1-openssl.pc: command not found
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating include/xmlsec/openssl/Makefile
config.status: error: cannot find input file: `src/openssl/Makefile.in'
make: *** [/c/Users/et5341/source/windows/cisbuild/thirdparty.mk:77: configure] Error 1
```

Sorry, I am no expert on these, could you tell me what's the correct way to regenerate those `configure` files? :D

**Comment by lsh123** [comment]
Run `autogen.sh <configure options>` from the root (https://github.com/lsh123/xmlsec/blob/master/autogen.sh)

**Comment by totszwai** [comment]
Is fine, no rush, I will test/check it locally first and report back. Thank you!

**Comment by totszwai** [comment]
I confirm that everything is working fine on my side now, thank you @lsh123

**Comment by lsh123** [comment]
Thanks, @totszwai !

I published new 1.2.38 release that includes this fix:

https://github.com/lsh123/xmlsec/discussions/687
